extensions/omni_txn/omni_txn.c

```diff
@@ -0,0 +1,81 @@
+// clang-format off
+#include <postgres.h>
+#include <fmgr.h>
+// clang-format on
+#include <executor/spi.h>
+#include <miscadmin.h>
+#include <utils/builtins.h>
+#include <utils/snapmgr.h>
+#if PG_MAJORVERSION_NUM < 15
+#include <access/xact.h>
+#endif
+
+PG_MODULE_MAGIC;
+
+static int32 retry_attempts = 0;
+
+PG_FUNCTION_INFO_V1(retry);
+
+Datum retry(PG_FUNCTION_ARGS) {
+  if (PG_ARGISNULL(0)) {
+    ereport(ERROR, errmsg("transaction statements argument is required"));
+  }
+  if (IsTransactionBlock()) {
```

Comment:
There is still a case where it doesn't work: 

When we're not connected over a backend going through the wire protocol, but instead using something like a background worker, this check may not be doing anything.

We may very well be in an ongoing transaction but `IsTransactionBlock()` will return `false`.

So far I found that starting an implicit transaction (`BeginImplicitTransactionBlock`) around the execution makes this check effective, but it also makes it impossible to use this function even if we just wanted to simply call the procedure.

I simulated the change here:

```diff
diff --git a/extensions/omni_httpd/http_worker.c b/extensions/omni_httpd/http_worker.c
index ca473f1..de46842 100644
--- a/extensions/omni_httpd/http_worker.c
+++ b/extensions/omni_httpd/http_worker.c
@@ -652,6 +652,7 @@ static int handler(request_message_t *msg) {
   SetCurrentStatementStartTimestamp();
   StartTransactionCommand();
   PushActiveSnapshot(GetTransactionSnapshot());
+  BeginImplicitTransactionBlock();
   SPI_connect();
 
   bool succeeded = false;
@@ -894,6 +895,7 @@ static int handler(request_message_t *msg) {
 
 cleanup:
   SPI_finish();
+  EndImplicitTransactionBlock();
   PopActiveSnapshot();
   if (succeeded) {
     CommitTransactionCommand();
```

This could be better. Thinking of a better method to ensure that the transaction has _actually_ started.  🤔 


🤔

Reply:
A bigger problem here is that I don't yet know how to make this work in `omni_httpd` – currently, all handling goes through a transaction. It does work, but erroneously, because we don't begin a transaction block.

* Should omni_httpd (and future workers) handlers be configurable as retriable transactions?
* Should we support multi-transactional workloads?

Reply:
Upcoming omni_httpd supports multi-transactional handlers #550